// @lit-protocol/pkp-ethers/src/index.js
"use strict";
// ----- autogen:polyfills:start  -----
// This field will be automatically injected into the ./dist/packages/<package-name>/index.js file
// between the autogen:polyfills:start/end comments

try {
  globalThis.crypto = require('crypto').webcrypto;
} catch (e) {
  // swallow
}

// ----- autogen:polyfills:end  -----
Object.defineProperty(exports, "__esModule", { value: true });
exports.isEthRequest = exports.sendRawTransactionHandler = exports.sendTransactionHandler = exports.signTypedDataLegacy = exports.signTypedData = exports.signTypedDataHandler = exports.personalSignHandler = exports.signHandler = exports.validateAddressesMatch = exports.validateSignature = exports.getTypedDataVersionInfo = exports.ethRequestHandler = exports.methodHandlers = exports.PKPEthersWallet = void 0;
const pkp_ethers_1 = require("./lib/pkp-ethers");
Object.defineProperty(exports, "PKPEthersWallet", { enumerable: true, get: function () { return pkp_ethers_1.PKPEthersWallet; } });
const handler_1 = require("./lib/handler");
Object.defineProperty(exports, "methodHandlers", { enumerable: true, get: function () { return handler_1.methodHandlers; } });
Object.defineProperty(exports, "ethRequestHandler", { enumerable: true, get: function () { return handler_1.ethRequestHandler; } });
Object.defineProperty(exports, "getTypedDataVersionInfo", { enumerable: true, get: function () { return handler_1.getTypedDataVersionInfo; } });
Object.defineProperty(exports, "validateSignature", { enumerable: true, get: function () { return handler_1.validateSignature; } });
Object.defineProperty(exports, "validateAddressesMatch", { enumerable: true, get: function () { return handler_1.validateAddressesMatch; } });
Object.defineProperty(exports, "signHandler", { enumerable: true, get: function () { return handler_1.signHandler; } });
Object.defineProperty(exports, "personalSignHandler", { enumerable: true, get: function () { return handler_1.personalSignHandler; } });
Object.defineProperty(exports, "signTypedDataHandler", { enumerable: true, get: function () { return handler_1.signTypedDataHandler; } });
Object.defineProperty(exports, "signTypedData", { enumerable: true, get: function () { return handler_1.signTypedData; } });
Object.defineProperty(exports, "signTypedDataLegacy", { enumerable: true, get: function () { return handler_1.signTypedDataLegacy; } });
Object.defineProperty(exports, "sendTransactionHandler", { enumerable: true, get: function () { return handler_1.sendTransactionHandler; } });
Object.defineProperty(exports, "sendRawTransactionHandler", { enumerable: true, get: function () { return handler_1.sendRawTransactionHandler; } });
Object.defineProperty(exports, "isEthRequest", { enumerable: true, get: function () { return handler_1.isEthRequest; } });
