// @alchemy/aa-signers/dist/cjs/particle/signer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ParticleSigner = void 0;
const aa_core_1 = require("@alchemy/aa-core");
const auth_1 = require("@particle-network/auth");
const provider_1 = require("@particle-network/provider");
const viem_1 = require("viem");
const constants_js_1 = require("../constants.js");
class ParticleSigner {
    constructor(params) {
        Object.defineProperty(this, "inner", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "provider", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "signer", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "signerType", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: `${constants_js_1.signerTypePrefix}particle`
        });
        Object.defineProperty(this, "getAddress", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: async () => {
                if (!this.signer)
                    throw new Error("Not authenticated");
                const address = await this.signer.getAddress();
                if (address == null)
                    throw new Error("No address found");
                return address;
            }
        });
        Object.defineProperty(this, "signMessage", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: async (msg) => {
                if (!this.signer)
                    throw new Error("Not authenticated");
                return this.signer.signMessage(msg);
            }
        });
        Object.defineProperty(this, "signTypedData", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: async (params) => {
                if (!this.signer)
                    throw new Error("Not authenticated");
                return this.signer.signTypedData(params);
            }
        });
        Object.defineProperty(this, "authenticate", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: async (params = {
                loginOptions: {},
                login: async (loginOptions) => {
                    this.inner.auth.login(loginOptions);
                },
            }) => {
                if (this.inner == null)
                    throw new Error("No provider found");
                await params.login(params.loginOptions);
                this.signer = new aa_core_1.WalletClientSigner((0, viem_1.createWalletClient)({
                    transport: (0, viem_1.custom)(this.provider),
                }), this.signerType);
                const userInfo = this.inner.auth.getUserInfo();
                if (userInfo == null)
                    throw new Error("No user info found");
                return userInfo;
            }
        });
        Object.defineProperty(this, "getAuthDetails", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: async () => {
                if (!this.signer)
                    throw new Error("Not authenticated");
                const userInfo = this.inner.auth.getUserInfo();
                if (userInfo == null)
                    throw new Error("No user info found");
                return userInfo;
            }
        });
        if ("inner" in params) {
            this.inner = params.inner;
            this.provider =
                params.provider != null
                    ? params.provider
                    : new provider_1.ParticleProvider(this.inner.auth);
            if (this.inner.auth.isLogin()) {
                this.signer = new aa_core_1.WalletClientSigner((0, viem_1.createWalletClient)({
                    transport: (0, viem_1.custom)(this.provider),
                }), this.signerType);
            }
            return;
        }
        this.inner = new auth_1.ParticleNetwork(params);
        this.provider = new provider_1.ParticleProvider(this.inner.auth);
        if (this.inner.auth.isLogin()) {
            this.signer = new aa_core_1.WalletClientSigner((0, viem_1.createWalletClient)({
                transport: (0, viem_1.custom)(this.provider),
            }), this.signerType);
        }
    }
}
exports.ParticleSigner = ParticleSigner;
