// @alchemy/aa-signers/dist/esm/passport/signer.js
import { WalletClientSigner, } from "@alchemy/aa-core";
import { signerTypePrefix } from "../constants.js";
import { Passport } from "@0xpass/passport";
import { createPassportClient } from "@0xpass/passport-viem-alchemy";
import { http } from "viem";
export class PassportSigner {
    constructor(params) {
        Object.defineProperty(this, "inner", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "signer", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "authDetails", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "signerType", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: `${signerTypePrefix}passport`
        });
        Object.defineProperty(this, "getAddress", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: async () => {
                if (!this.inner || !this.signer) {
                    throw new Error("Not authenticated");
                }
                return this.signer.getAddress();
            }
        });
        Object.defineProperty(this, "signMessage", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: async (msg) => {
                if (!this.inner || !this.signer) {
                    throw new Error("Not authenticated");
                }
                return this.signer.signMessage(msg);
            }
        });
        Object.defineProperty(this, "signTypedData", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: async (params) => {
                if (!this.inner || !this.signer) {
                    throw new Error("Not authenticated");
                }
                return this.signer.signTypedData(params);
            }
        });
        Object.defineProperty(this, "authenticate", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: async (params) => {
                if (this.inner == null)
                    throw new Error("No provider found");
                await this.inner.setupEncryption();
                const [authenticatedHeaders, address] = await this.inner.authenticate({
                    username: params.username,
                    userDisplayName: params.userDisplayName,
                });
                const fallbackProvider = http(params.fallbackProvider);
                const client = params.endpoint
                    ? createPassportClient(authenticatedHeaders, fallbackProvider, params.chain, params.endpoint)
                    : createPassportClient(authenticatedHeaders, fallbackProvider, params.chain);
                this.signer = new WalletClientSigner(client, this.signerType);
                this.authDetails = {
                    authenticatedHeaders: authenticatedHeaders,
                    addresses: [address],
                };
                return this.authDetails;
            }
        });
        Object.defineProperty(this, "getAuthDetails", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: async () => {
                if (!this.inner || !this.signer || !this.authDetails) {
                    throw new Error("Not authenticated");
                }
                return this.authDetails;
            }
        });
        if ("inner" in params) {
            this.inner = params.inner;
        }
        else {
            this.inner = new Passport(params);
        }
    }
}
