// @usecapsule/web-sdk/dist/core/definitions.js
/* eslint-disable */
global.Buffer = global.Buffer || require('buffer').Buffer;
export const is2FAEnabled = false;
export var Environment;
(function (Environment) {
    // Internal Environments
    Environment["DEV"] = "DEV";
    Environment["SANDBOX"] = "SANDBOX";
    Environment["BETA"] = "BETA";
    Environment["PROD"] = "PROD";
    // Customer-Facing Environments
    // NOTE: these resolve to the corresponding internal environments for convenience
    Environment["DEVELOPMENT"] = "BETA";
    Environment["PRODUCTION"] = "PROD";
})(Environment || (Environment = {}));
export function getPortalDomain(env) {
    switch (env) {
        case Environment.DEV:
            return 'localhost';
        case Environment.SANDBOX:
            return 'app.sandbox.usecapsule.com';
        case Environment.BETA:
            return 'app.beta.usecapsule.com';
        case Environment.PROD:
            return 'app.usecapsule.com';
        default:
            throw new Error(`env: ${env} not supported`);
    }
}
export function getPortalBaseURL({ env }, useLocalIp) {
    const domain = getPortalDomain(env);
    if (env === Environment.DEV) {
        if (useLocalIp) {
            return `http://127.0.0.1:3003`;
        }
        return `http://${domain}:3003`;
    }
    return `https://${domain}`;
}
